{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "import json\n",
    "import numpy as np\n",
    "import requests\n",
    "import sys\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "###Create Tools for code\n",
    "def append_file_to_list(file):\n",
    "    list = []\n",
    "    f = open(file, 'r')\n",
    "    for line in f:\n",
    "        line = line.strip('\\n')\n",
    "        line = line.split('\\t')\n",
    "        list.append(line)\n",
    "    return list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Pull in Data from JSON\n",
    "CIViC = requests.get('https://civic.genome.wustl.edu/api/variants?count=1000000').json()['records']\n",
    "DoCM = requests.get('http://docm.genome.wustl.edu/api/v1/variants?count=1000000').json()\n",
    "variants_capture = requests.get('https://civic.genome.wustl.edu/api/panels/captureseq/qualifying_variants?minimum_score=0').json()['records']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "#open error file for deep learning\n",
    "error_data = append_file_to_list('deep_learning/errors_dl.tsv')\n",
    "error_pd = pd.read_csv('deep_learning/errors_dl.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Open SOID files\n",
    "eligible_SOIDs = []\n",
    "f = open('SOID_terms.txt', 'r')\n",
    "for line in f:\n",
    "    line = line.strip('\\n')\n",
    "    line = line.strip()\n",
    "    eligible_SOIDs.append(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Discrepencies:  4378\n",
      "Number of False Positives:  1659\n",
      "Number of False Negatives:  2719\n"
     ]
    }
   ],
   "source": [
    "print('Number of Discrepencies: ', len(error_pd[(error_pd.error=='False Negative') | (error_pd.error=='False Positive')]))\n",
    "print('Number of False Positives: ', len(error_pd[(error_pd.error=='False Positive')]))\n",
    "print('Number of False Negatives: ', len(error_pd[(error_pd.error=='False Negative')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "false_neg = []\n",
    "false_pos = []\n",
    "true_pos = []\n",
    "for item in error_data:\n",
    "    if item[2] == 'False Negative':\n",
    "        false_neg.append([item[7], item[8], item[9], item[10], item[11], item[12], item[5], item[0], item[1]])\n",
    "    if item[2] == 'False Positive':\n",
    "        false_pos.append([item[7], item[8], item[9], item[10], item[11], item[12], item[5], item[0], item[1]])\n",
    "    if item[2] == 'True Positive':\n",
    "        true_pos.append([item[7], item[8], item[9], item[10], item[11], item[12], item[5], item[0], item[1]])\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of DoCM Annotations: 1364\n",
      "10\n",
      "14\n",
      "226\n"
     ]
    }
   ],
   "source": [
    "#pull docm coordinates for overlap\n",
    "DoCM_coordinates = []\n",
    "for item in DoCM:\n",
    "    variant = item['variant']\n",
    "    start = item['start']\n",
    "    stop = item['stop']\n",
    "    chrom = item['chr']\n",
    "    reference = item['read']\n",
    "    diseases = item['diseases']\n",
    "    name = item['gene']\n",
    "    citation = len(item['pubmed_sources'])\n",
    "    DoCM_coordinates.append([chrom, start, stop, reference, variant, name, citation])\n",
    "\n",
    "DoCM_clinical_variants_FN = []\n",
    "for item in false_neg:\n",
    "    for exon in DoCM_coordinates:\n",
    "        if str(item[0]) == str(exon[0]) and int(item[1]) == int(exon[1]) and int(item[2]) == int(exon[2]) and str(item[3]) == str(exon[3]) and str(item[4]) == str(exon[4]):\n",
    "            DoCM_clinical_variants_FN.append([item[0], item[1], item[2], item[3], item[4], exon[5], item[5], item[6], item[7], item[8], exon[6]])\n",
    "\n",
    "DoCM_clinical_variants_FP = []            \n",
    "for item in false_pos:\n",
    "    for exon in DoCM_coordinates:\n",
    "        if str(item[0]) == str(exon[0]) and int(item[1]) == int(exon[1]) and int(item[2]) == int(exon[2]) and str(item[3]) == str(exon[3]) and str(item[4]) == str(exon[4]):\n",
    "            DoCM_clinical_variants_FP.append([item[0], item[1], item[2], item[3], item[4], exon[5], item[5], item[6], item[7], item[8], exon[6]])\n",
    "            \n",
    "DoCM_clinical_variants_TP = []            \n",
    "for item in true_pos:\n",
    "    for exon in DoCM_coordinates:\n",
    "        if str(item[0]) == str(exon[0]) and int(item[1]) == int(exon[1]) and int(item[2]) == int(exon[2]) and str(item[3]) == str(exon[3]) and str(item[4]) == str(exon[4]):\n",
    "            DoCM_clinical_variants_TP.append([item[0], item[1], item[2], item[3], item[4], exon[5], item[5], item[6], item[7], item[8], exon[6]])\n",
    "            \n",
    "print('Number of DoCM Annotations:', len(DoCM))\n",
    "print(len(DoCM_clinical_variants_FN))\n",
    "print(len(DoCM_clinical_variants_FP))\n",
    "print(len(DoCM_clinical_variants_TP))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False Negative Overlap with DoCM Annotations: 8\n",
      "False Positive Overlap with DoCM Annotations: 12\n",
      "True Positive Overlap with DoCM Annotations: 94\n",
      "Publications associated with True Positives is:  9846\n"
     ]
    }
   ],
   "source": [
    "#Print DoCM Variants to list\n",
    "final_FN = {}\n",
    "for item in DoCM_clinical_variants_FN:\n",
    "    code = str(item[0] + item[1] + item[2] + item[3] + item[4])\n",
    "    if code in final_FN.keys():\n",
    "        final_FN[code][10] += item[10] \n",
    "    if code not in final_FN.keys():\n",
    "        final_FN[code] = ''\n",
    "        final_FN[code] = item\n",
    "print('False Negative Overlap with DoCM Annotations:', len(final_FN))\n",
    "\n",
    "#Print DoCM Variants to list\n",
    "final_FP = {}\n",
    "for item in DoCM_clinical_variants_FP:\n",
    "    code = str(item[0] + item[1] + item[2] + item[3] + item[4])\n",
    "    if code in final_FP.keys():\n",
    "        final_FP[code][10] += item[10] \n",
    "    if code not in final_FP.keys():\n",
    "        final_FP[code] = ''\n",
    "        final_FP[code] = item\n",
    "print('False Positive Overlap with DoCM Annotations:', len(final_FP))\n",
    "\n",
    "#Print DoCM Variants to list\n",
    "final_TP = {}\n",
    "for item in DoCM_clinical_variants_TP:\n",
    "    code = str(item[0] + item[1] + item[2] + item[3] + item[4])\n",
    "    if code in final_TP.keys():\n",
    "        final_TP[code][10] += item[10] \n",
    "    if code not in final_TP.keys():\n",
    "        final_TP[code] = ''\n",
    "        final_TP[code] = item\n",
    "print('True Positive Overlap with DoCM Annotations:', len(final_TP))\n",
    "pubs_TP = 0\n",
    "for k,v in final_TP.items():\n",
    "    pubs_TP += int(v[10])\n",
    "print('Publications associated with True Positives is: ',pubs_TP)\n",
    "\n",
    "\n",
    "clinical = open('deep_learning/DoCM_variants_DL_FN.txt', 'w')\n",
    "clinical.write('DoCM Annotation Overlap with False Negatives' + '\\n')\n",
    "clinical.write('Chr' + '\\t' + 'Start' + '\\t'+ 'Stop' + '\\t' + 'Ref' + '\\t' + 'Var' + '\\t' + 'Gene' + '\\t' + 'Disease' + '\\t' + 'Probability' + '\\t' + 'MR Call' + '\\t' + 'Classifier' + '\\t' + 'Publications'+ '\\n')\n",
    "for k,v in final_FN.items():\n",
    "    for bugger in v:\n",
    "        clinical.write(str(bugger) + '\\t')\n",
    "    clinical.write('\\n')\n",
    "clinical.close()\n",
    "\n",
    "clinical = open('deep_learning/DoCM_variants_DL_FP.txt', 'w')\n",
    "clinical.write('DoCM Annotation Overlap with False Positives' + '\\n')\n",
    "clinical.write('Chr' + '\\t' + 'Start' + '\\t'+ 'Stop' + '\\t' + 'Ref' + '\\t' + 'Var' + '\\t' + 'Gene' + '\\t' + 'Disease' + '\\t' + 'Probability' + '\\t' + 'MR Call' + '\\t' + 'Classifier' + '\\t' + 'Publications'+ '\\n')\n",
    "for k,v in final_FP.items():\n",
    "    for bugger in v:\n",
    "        clinical.write(str(bugger) + '\\t')\n",
    "    clinical.write('\\n')\n",
    "clinical.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Eligible Loci within CIViC: 425\n",
      "True Positive Sensitivity =  772\n",
      "True Positive Resistance =  547\n",
      "True Positive Prognosis =  176\n",
      "True Positive Diagnostic =  50\n",
      "True Positive Predisposition =  307\n"
     ]
    }
   ],
   "source": [
    "#Pull items from CIViC\n",
    "score = {'A':5, 'B':4, 'C':3, 'D':2, 'E':1}\n",
    "\n",
    "capture_sequence_probes = [] #create empty list for capture sequence probes\n",
    "for k in range(0, len(variants_capture)): #iterate through API and pull all eligible variants\n",
    "    counts = [0,0,0,0,0]\n",
    "    gene = variants_capture[k]['entrez_name']  #Call Gene name\n",
    "    variant = variants_capture[k]['name'] #call variant\n",
    "    soid = variants_capture[k]['variant_types'][0]['so_id'] #call soid\n",
    "    variant_type = variants_capture[k]['variant_types'][0]['name'] #call variant type\n",
    "    evidence = variants_capture[k]['evidence_items'] #pull evidence items\n",
    "    evidence_statements = len(variants_capture[k]['evidence_items']) #pull number of evidence statements\n",
    "    chrom = variants_capture[k]['coordinates']['chromosome'] #call chrom\n",
    "    start = variants_capture[k]['coordinates']['start'] #call start\n",
    "    stop = variants_capture[k]['coordinates']['stop'] #call stop\n",
    "    ref = variants_capture[k]['coordinates']['reference_bases']\n",
    "    var = variants_capture[k]['coordinates']['variant_bases']\n",
    "    evidence_type = [] #set list for evidence types\n",
    "    evidence_scores = [] #set list for evidence scores\n",
    "    top_evidences = [] #set list for top evidence level\n",
    "    for item in evidence: #iterate through the evidence items\n",
    "        if item['evidence_type'] not in evidence_type: #see if the evidence type is already there\n",
    "            evidence_type.append(item['evidence_type']) #if it is not append it\n",
    "        if item['evidence_type'] == 'Predictive':\n",
    "            if item['clinical_significance'] == 'Sensitive' or item['clinical_significance'] == 'Sensitivity':\n",
    "                counts[0] += 1\n",
    "            elif item['clinical_significance'] == 'Resistance or Non-Response': # or 'Adverse' in item['clinical_significance']:\n",
    "                counts[1] += 1\n",
    "            elif item['clinical_significance'] == 'Adverse Response': # or 'Adverse' in item['clinical_significance']:\n",
    "                counts[1] += 1\n",
    "        elif item['evidence_type'] == 'Prognostic':\n",
    "            counts[2] += 1\n",
    "        elif item['evidence_type'] == 'Diagnostic':\n",
    "            counts[3] += 1\n",
    "        elif item['evidence_type'] == 'Predisposing':\n",
    "            counts[4] += 1\n",
    "        trust_rating = int(item['rating'] or 0) #make the trust rating either what is listed or 0\n",
    "        evidence_level = int(score[item['evidence_level']]) #make the evidence level the value from the score dictionary\n",
    "        evidence_scores.append(evidence_level * trust_rating) #calculate the Evidence Score\n",
    "        if item['evidence_level'] != '[]': #find the evidence levels that are not blank\n",
    "            top_evidences.append(item['evidence_level'].strip()) #add to the list\n",
    "    #pull the maximum evidence level\n",
    "    if 'A' in top_evidences:\n",
    "        top_evidence = 'A'\n",
    "    elif 'B' in top_evidences:\n",
    "        top_evidence = 'B'\n",
    "    elif 'C' in top_evidences:\n",
    "        top_evidence = 'C'\n",
    "    elif 'D' in top_evidences:\n",
    "        top_evidence = 'D'\n",
    "    else:\n",
    "        top_evidence = 'E'\n",
    "    evidence_score = sum(evidence_scores) #sum the evidence scores to get a CIVic Score\n",
    "    evidence_types = ', '.join(evidence_type) #format the evidence types\n",
    "    if soid in eligible_SOIDs:\n",
    "        capture_sequence_probes.append([chrom, start,stop, ref, var, gene, soid, variant_type, variant, top_evidence, evidence_types, evidence_statements, evidence_score, counts])\n",
    "\n",
    "print('Number of Eligible Loci within CIViC:', len(capture_sequence_probes))\n",
    "\n",
    "sen = 0\n",
    "res = 0\n",
    "pro = 0\n",
    "dia = 0\n",
    "pre = 0\n",
    "\n",
    "for item in capture_sequence_probes:\n",
    "    sen += int(item[13][0])\n",
    "    res += int(item[13][1])\n",
    "    pro += int(item[13][2])\n",
    "    dia += int(item[13][3])\n",
    "    pre += int(item[13][4])\n",
    "        \n",
    "print('True Positive Sensitivity = ', sen) \n",
    "print('True Positive Resistance = ', res)   \n",
    "print('True Positive Prognosis = ', pro)   \n",
    "print('True Positive Diagnostic = ', dia)   \n",
    "print('True Positive Predisposition = ', pre)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Exact Unique Overlaps with CIViC: 3\n",
      "Number of Total Unique Overlaps with CIViC: 51\n",
      "Overlap with CIViC Variants: 40\n"
     ]
    }
   ],
   "source": [
    "#Elucidate False Negatives that have exact and direct overlap with CIViC\n",
    "\n",
    "capture_overlap_unique_exact_FN = []\n",
    "capture_overlap_unique_about_FN = []\n",
    "\n",
    "for item in false_neg:\n",
    "    for exon in capture_sequence_probes:\n",
    "        code = [item[0], item[1], item[2], item[3], item[4], exon[5], item[5], item[6], item[7], item[8], exon[8], exon[9], exon[10], exon[11], exon[12], exon[13]]\n",
    "        if str(item[0]) == str(exon[0]) and int(item[1]) == int(exon[1])and int(item[2]) == int(exon[2]) and str(item[3]) == str(exon[3]) and str(item[4]) == str(exon[4]) and code not in capture_overlap_unique_exact_FN:\n",
    "            capture_overlap_unique_exact_FN.append(code)\n",
    "        if str(item[0]) == str(exon[0]) and int(item[1]) >= int(exon[1]) and int(item[2]) <= int(exon[2]) and code not in capture_overlap_unique_about_FN:\n",
    "            capture_overlap_unique_about_FN.append(code)\n",
    "\n",
    "print('Number of Exact Unique Overlaps with CIViC:', len(capture_overlap_unique_exact_FN))\n",
    "print('Number of Total Unique Overlaps with CIViC:', len(capture_overlap_unique_about_FN))\n",
    "\n",
    "final_FN = {}\n",
    "\n",
    "for item in capture_overlap_unique_about_FN:\n",
    "    code = str(item[0] + item[1] + item[2] + item[3] + item[4] + item[7])\n",
    "    if code in final_FN.keys():\n",
    "        variant = final_FN[code][10].split(',')\n",
    "        type_thing =  final_FN[code][12].split(',')\n",
    "        if item[10] not in variant:\n",
    "            final_FN[code][10] += ', '\n",
    "            final_FN[code][10] += item[10]\n",
    "        if item[12] not in type_thing:\n",
    "            final_FN[code][12] += ', '\n",
    "            final_FN[code][12] += item[12]\n",
    "        final_FN[code][13] += item[13]\n",
    "        final_FN[code][14] += item[14]\n",
    "    if code not in final_FN.keys():\n",
    "        final_FN[code] = ''\n",
    "        final_FN[code] = item\n",
    "\n",
    "print('Overlap with CIViC Variants:', len(final_FN))\n",
    "\n",
    "clinical = open('deep_learning/clinical_variants_False_Negatives.txt', 'w')\n",
    "clinical.write('False Negative Overlap with CIViC' + '\\n')\n",
    "clinical.write(\"Chr\" + '\\t' + \"Start\" + '\\t' + \"Stop\" + '\\t' + \"Ref\" + '\\t' \n",
    "               + \"Var\" + '\\t' + 'Gene' + '\\t' + \"Disease\" + '\\t' + \"Confidence\" \n",
    "               + '\\t' + \"MR Call\" + '\\t' + \"Classifier Call\" + '\\t' + \"Variant\" \n",
    "               + '\\t' + \"Top Evidence\" + '\\t' + \"Evidence Type\" + '\\t' \n",
    "               + \"Evidence Items\" + '\\t' + \"CIViC Score\" + '\\t' + \"Sensitivity\"\n",
    "               + '\\t' + \"Resistance\" + '\\t' + \"Prognosis\" + '\\t' + \"Diagnosis\"\n",
    "               + '\\t' + \"Predisposing\" + \"\\n\")\n",
    "for k,v in final_FN.items():\n",
    "    for item in range(15):\n",
    "        clinical.write(str(v[item]) + '\\t')\n",
    "    for item in v[15]:\n",
    "        clinical.write(str(item) + '\\t')\n",
    "    clinical.write('\\n')\n",
    "clinical.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Exact Unique Overlaps with CIViC for False Positives: 5\n",
      "Number of Total Unique Overlaps with CIViC for False Positives: 65\n",
      "False Positive Overlap with CIViC Variants: 53\n"
     ]
    }
   ],
   "source": [
    "#Elucidate False Negatives that have exact and direct overlap with CIViC\n",
    "\n",
    "capture_overlap_unique_exact_FP = []\n",
    "capture_overlap_unique_about_FP = []\n",
    "\n",
    "for item in false_pos:\n",
    "    for exon in capture_sequence_probes:\n",
    "        code = [item[0], item[1], item[2], item[3], item[4], exon[5], item[5], item[6], item[7], item[8], exon[8], exon[9], exon[10], exon[11], exon[12], exon[13]]\n",
    "        if str(item[0]) == str(exon[0]) and int(item[1]) == int(exon[1]) and int(item[2]) == int(exon[2]) and str(item[3]) == str(exon[3]) and str(item[4]) == str(exon[4]) and code not in capture_overlap_unique_exact_FP:\n",
    "            capture_overlap_unique_exact_FP.append(code)\n",
    "        if str(item[0]) == str(exon[0]) and int(item[1]) >= int(exon[1]) and int(item[2]) <= int(exon[2]) and code not in capture_overlap_unique_about_FP:\n",
    "            capture_overlap_unique_about_FP.append(code)\n",
    "\n",
    "print('Number of Exact Unique Overlaps with CIViC for False Positives:', len(capture_overlap_unique_exact_FP))\n",
    "print('Number of Total Unique Overlaps with CIViC for False Positives:', len(capture_overlap_unique_about_FP))\n",
    "\n",
    "final_FP = {}\n",
    "\n",
    "for item in capture_overlap_unique_about_FP:\n",
    "    code = str(item[0] + item[1] + item[2] + item[3] + item[4] + item[7])\n",
    "    if code in final_FP.keys():\n",
    "        variant = final_FP[code][10].split(',')\n",
    "        type_thing =  final_FP[code][12].split(',')\n",
    "        if item[10] not in variant:\n",
    "            final_FP[code][10] += ', '\n",
    "            final_FP[code][10] += item[10]\n",
    "        if item[12] not in type_thing:\n",
    "            final_FP[code][12] += ', '\n",
    "            final_FP[code][12] += item[12]\n",
    "        final_FP[code][13] += item[13]\n",
    "        final_FP[code][14] += item[14]\n",
    "    if code not in final_FP.keys():\n",
    "        final_FP[code] = ''\n",
    "        final_FP[code] = item\n",
    "\n",
    "print('False Positive Overlap with CIViC Variants:', len(final_FP))\n",
    "\n",
    "clinical = open('deep_learning/clinical_variants_False_Positive.txt', 'w')\n",
    "clinical.write('False Positive Overlap with CIViC' + '\\n')\n",
    "clinical.write(\"Chr\" + '\\t' + \"Start\" + '\\t' + \"Stop\" + '\\t' + \"Ref\" + '\\t' \n",
    "               + \"Var\" + '\\t' + 'Gene' + '\\t' + \"Disease\" + '\\t' + \"Confidence\" \n",
    "               + '\\t' + \"MR Call\" + '\\t' + \"Classifier Call\" + '\\t' + \"Variant\" \n",
    "               + '\\t' + \"Top Evidence\" + '\\t' + \"Evidence Type\" + '\\t' \n",
    "               + \"Evidence Items\" + '\\t' + \"CIViC Score\" + '\\t' + \"Sensitivity\"\n",
    "               + '\\t' + \"Resistance\" + '\\t' + \"Prognosis\" + '\\t' + \"Diagnosis\"\n",
    "               + '\\t' + \"Predisposing\" + \"\\n\")\n",
    "for k,v in final_FP.items():\n",
    "    for item in range(15):\n",
    "        clinical.write(str(v[item]) + '\\t')\n",
    "    for item in v[15]:\n",
    "        clinical.write(str(item) + '\\t')\n",
    "    clinical.write('\\n')\n",
    "clinical.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Determine Total Number of Variants that overlap with CIViC for all calls\n",
    "error_data_bed = {}\n",
    "for idx,item in enumerate(error_data):\n",
    "    error_data_bed[idx] = [item, [0,0,0,0,0]]\n",
    "    for exon in capture_sequence_probes:\n",
    "        if str(item[7]) == str(exon[0]) and int(item[8]) >= int(exon[1]) and int(item[9]) <= int(exon[2]):\n",
    "            error_data_bed[idx][1][0] += exon[13][0]\n",
    "            error_data_bed[idx][1][1] += exon[13][1]\n",
    "            error_data_bed[idx][1][2] += exon[13][2]\n",
    "            error_data_bed[idx][1][3] += exon[13][3]\n",
    "            error_data_bed[idx][1][4] += exon[13][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "both = 0\n",
    "manual_review_somatic = 0\n",
    "deep_learning_somatic = 0\n",
    "\n",
    "both_list = {}\n",
    "manual_review_somatic_list = {}\n",
    "deep_learning_somatic_list = {}\n",
    "\n",
    "variants = {}\n",
    "for item in range(len(error_data) - 1):\n",
    "    if item not in variants.keys():\n",
    "        variants[item] = error_data[item + 1]\n",
    "\n",
    "for k,v in error_data_bed.items():\n",
    "    if v[0][0] == 's' and v[0][1] == 's':\n",
    "        both += 1\n",
    "        both_list[k] = v\n",
    "    elif v[0][0] == 's' and v[0][1] != 's':\n",
    "        manual_review_somatic += 1\n",
    "        manual_review_somatic_list[k] = v\n",
    "    elif v[0][0] != 's' and v[0][1] == 's':\n",
    "        deep_learning_somatic += 1\n",
    "        deep_learning_somatic_list[k] = v\n",
    "\n",
    "print('Variants in Both: ', both)\n",
    "print('Variants in Manual Review Only: ', manual_review_somatic)\n",
    "print('Variants in Deep Learning Only: ', deep_learning_somatic)\n",
    "print()\n",
    "\n",
    "mr_counts = 0\n",
    "deep_counts = 0\n",
    "both_counts = 0\n",
    "\n",
    "mr_clinical = [0,0,0,0,0]\n",
    "dl_clinical = [0,0,0,0,0]\n",
    "both_clinical = [0,0,0,0,0]\n",
    "\n",
    "mr_clinical_list = {}\n",
    "dl_clinical_list = {}\n",
    "both_clinical_list = {}\n",
    "\n",
    "for k,v in both_list.items():\n",
    "    if v[1] != [0,0,0,0,0]:\n",
    "        both_counts += 1\n",
    "        both_clinical_list[k] = v\n",
    "        both_clinical[0] += v[1][0]\n",
    "        both_clinical[1] += v[1][1]\n",
    "        both_clinical[2] += v[1][2]\n",
    "        both_clinical[3] += v[1][3]\n",
    "        both_clinical[4] += v[1][4]\n",
    "\n",
    "for k,v in manual_review_somatic_list.items():\n",
    "    if v[1] != [0,0,0,0,0]:\n",
    "        mr_counts += 1\n",
    "        mr_clinical_list[k] = v\n",
    "        mr_clinical[0] += v[1][0]\n",
    "        mr_clinical[1] += v[1][1]\n",
    "        mr_clinical[2] += v[1][2]\n",
    "        mr_clinical[3] += v[1][3]\n",
    "        mr_clinical[4] += v[1][4]\n",
    "\n",
    "for k,v in deep_learning_somatic_list.items():\n",
    "    if v[1] != [0,0,0,0,0]:\n",
    "        deep_counts += 1\n",
    "        dl_clinical_list[k] = v\n",
    "        dl_clinical[0] += v[1][0]\n",
    "        dl_clinical[1] += v[1][1]\n",
    "        dl_clinical[2] += v[1][2]\n",
    "        dl_clinical[3] += v[1][3]\n",
    "        dl_clinical[4] += v[1][4]\n",
    "        \n",
    "print('Clinical Variants in Both: ', both_counts)\n",
    "print('Clinical Variants in Manual Review Only: ', mr_counts)\n",
    "print('Clinical Variants in Deep Learning Only: ', deep_counts)\n",
    "print()\n",
    "print('Clinical Variants in Both: ', both_clinical)\n",
    "print('Clinical Variants in Manual Review Only: ', mr_clinical)\n",
    "print('Clinical Variants in Deep Learning Only: ', dl_clinical)\n",
    "\n",
    "avg = []\n",
    "for k,v in mr_clinical_list.items():\n",
    "    avg.append(float(v[0][4]))\n",
    "    \n",
    "file = open('manual_review_to_review.tsv', 'w')\n",
    "for k,v in mr_clinical_list.items():\n",
    "    file.write(str(v[0][0]) + '\\t' + str(v[0][6]) + '\\t' + str(v[0][7]) + '\\t' + str(v[0][8]) + '\\t' +  str(v[0][9]) + '\\n')\n",
    "file.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
